### ISSUE TITLE
[Placeholder] Intel TDX enablement

### ISSUE THREAD

**Issue by LeiZhou-97** [kubevirt/kubevirt#9669]
This is just a placeholder issue for tracking the development progress of Intel TDX enaling in KubeVirt.

Currently, due to the Intel TDX kernel, qemu and libvirt upstream tasks are working in progress, we create a [intel/kubevirt-tdx](https://github.com/intel/kubevirt-tdx/tree/tdx-support) repo for development, which gives Intel TDX target customers a way to build the Intel TDX workable KubeVirt images. But building these images depends on the Linux Software Stack for Intel TDX, you can get it from [tdx-tools](https://github.com/intel/tdx-tools).

Our intel/kubevirt-tdx is based on the CentOS Stream 8 code base. The Linux Software Stack for Intel TDX does not support CentOS Stream 9 (KubeVirt base image) yet. We will provide support later.

## Implementation

- [x] Basic Intel TDX functional support 

- need rebase to the main branch after supporting CentOS Stream 9

- [x] Support Intel TDX Attestation
- [ ] Support Live Migration
- [x] Expose Key number of Intel TDX for TD guest management
- [ ] Support E2E Attestation use case FDE (Full Disk Encryption)

## Testing
- [ ] Add upstream CI for running functional tests with attestation
- [ ]  Create the test image

## Documentation
- [ ]  Initial design proposal
- [ ]  Initial usage guide

**Comment by alicefr** [comment]
@LeiZhou-97 do you plan on the long term to integrate the changes into kubevirt main repo? If so, please consider in your plan the upstream CI. We added a SEV AMD machine and there are still pr ongoing for running functional tests with attestation. It would be possible to have a similar approach for TDX.

**Comment by LeiZhou-97** [comment]
@alicefr Thanks for your suggestion. I'll add it to my task list.

**Comment by alicefr** [comment]
> Hi, @alicefr I am sorry for taking so long to continue to have some new updates about Intel TDX!
> 
> Since Intel TDX is a very complex technology, upstream is still WIP, but in order to make TDX available to end-users earlier, there is already a TDX 1.0 repo in the CentOS Stream repositories. https://linux-mirrors.fnal.gov/linux/centos-stream/SIGs/9-stream/virt/x86_64/tdx-devel/
> 
> So I would like to ask if the code about TDX VM enabling can be upstreamed based on a TDX mid-stream PKGs? This may require the user to build the image themselves or we can provide pre-built TDX-enabled images at each release directly. (because the TDX stacks do not use the standard qemu/libvirt) Or can I keep a TDX branch in the kubevirt repo for end-users to use?
> 
> Now, we’ve finished the basic TDX enabling based on the v1.0.0 release in the intel org repo. https://github.com/intel/kubevirt-tdx/tree/tdx-1.5 https://github.com/intel/kubevirt-tdx/wiki
> 
> Appreciate for your comments!

@LeiZhou-97 please send an email to kubevirt mailing list, this is something that needs to be discussed and agreed by the KubeVirt community and approvers. 

However, I find it difficult to have 2 libvirt and QEMU versions. We cannot test if something breaks on TDX if the images aren't the same as upstream. From my side, I think getting the changes upstreamed in libvirt and QEMU is a must before starting the work in KubeVirt.

**Comment by vladikr** [comment]
@MatiasVara would you like to be assigned to this?

**Comment by MatiasVara** [comment]
> [@MatiasVara](https://github.com/MatiasVara) would you like to be assigned to this? [@alicefr](https://github.com/alicefr) fyi.

yes, you can.

**Comment by mythi** [comment]
> > [@MatiasVara](https://github.com/MatiasVara) would you like to be assigned to this? [@alicefr](https://github.com/alicefr) fyi.
> 
> yes, you can.

@MatiasVara are you planning to work on a PR for this?

**Comment by MatiasVara** [comment]
> > > [@MatiasVara](https://github.com/MatiasVara) would you like to be assigned to this? [@alicefr](https://github.com/alicefr) fyi.
> > 
> > 
> > yes, you can.
> 
> [@MatiasVara](https://github.com/MatiasVara) are you planning to work on a PR for this?

Hello, yes, I am. I am currently working on a PoC but I do not have it working yet.

**Comment by mythi** [comment]
@MatiasVara great! I can help with any TDX specific questions you may have. Also feel free to tag me in your PR.

**Comment by MatiasVara** [comment]
> [@MatiasVara](https://github.com/MatiasVara) great! I can help with any TDX specific questions you may have. Also feel free to tag me in your PR.

Cool! I am currently struggling with the attestation. I setup the virt-launcher with the right QEMU and I have included the qsg service. Also, I setup a device plugin to expose `/dev/sgx-*`. However, I am getting the following issue:
```
[QCNL] Encountered CURL error: (77) Problem with the SSL CA cert (path? access rights?) 

[QPL] Failed to get quote config. Error code is 0xb003

[get_platform_quote_cert_data ../td_ql_logic.cpp:302] Error returned from the p_sgx_get_quote_config API. 0xe019
tee_att_init_quote return 0x11001
tee_att_get_quote_size return 0x1100f
resp_size is 0
About to shutdown and close socket
erased a connection, now [0]
``` 
I think something else is missing the virt-launcher pod. That made me think that qgs shall be in a different container and only try to share the unix socket with the virt-launcher although I do not how to do it yet.

**Comment by mythi** [comment]
> That made me think that qgs shall be in a different container and only share the unix socket with the virt-launcher although I do not how to do it yet.

Do you have QGS running in a container? You will also need to [configure it to point to a proper PCCS service](https://cc-enabling.trustedservices.intel.com/intel-tdx-enabling-guide/05/host_os_setup/#configure-qcnl) in our cluster / infra.

**Comment by MatiasVara** [comment]
> > That made me think that qgs shall be in a different container and only share the unix socket with the virt-launcher although I do not how to do it yet.
> 
> Do you have QGS running in a container? You will also need to [configure it to point to a proper PCCS service](https://cc-enabling.trustedservices.intel.com/intel-tdx-enabling-guide/05/host_os_setup/#configure-qcnl) in our cluster / infra.

I think it is failing before contacting any PCCS service. I think some credential is missing since the virt-launcher is a very restricted.

**Comment by mythi** [comment]
> I think some credential is missing since the virt-launcher is very restricted and is not meant to run q binary like qgs there.

Running QGS in a dedicated DaemonSet with UDS (`port=0` config/param) made available via a `hostPath` mount should be enough and it does not need any extra privileges (other than getting `/dev/sgx_*` resources from the device plugin).  You might have to toggle `"use_secure_cert"` to `false` in your `sgx_default_qcnl.conf`

**Comment by victortoso** [comment]
@mythi do you know why the certificates is needed? Do you suggest disabling it now for development but we will need to properly setup it later on?

**Comment by mythi** [comment]
> [@mythi](https://github.com/mythi) do you know why the certificates is needed? Do you suggest disabling it now for development but we will need to properly setup it later on?

It's like any web service with properly signed TLS certificates. QGS talks to PCCS (collateral caching service) REST to get the quote signing collateral but PCCS is typically infra specific thing and something KubeVirt implementation itself does not need to worry about (KubeVirt can assume a properly configured QGS is running on each node). For testing and development (and since you're likely running PCCS of your own), disabling should be fine.

**Comment by berrange** [comment]
> QGS talks to PCCS (collateral caching service) REST to get the quote signing collateral but PCCS is typically infra specific thing

Note, Intel removed all the PCCS code from SGX in the DCP 1.22 release, so if you're using the open source SGX/DCAP code, QGS will be talking  directly to the public servers.

**Comment by zhencliu** [comment]
Hi, thanks for reopening this issue, I've been working on a PR(https://github.com/kubevirt/kubevirt/pull/14575) for the "Basic Intel TDX functional support(without attestation)", I marked the the PR as a draft for it's WIP, and will do more testings.

**Comment by MatiasVara** [comment]
> > I think some credential is missing since the virt-launcher is very restricted and is not meant to run q binary like qgs there.
> 
> Running QGS in a dedicated DaemonSet with UDS (`port=0` config/param) made available via a `hostPath` mount should be enough and it does not need any extra privileges (other than getting `/dev/sgx_*` resources from the device plugin). You might have to toggle `"use_secure_cert"` to `false` in your `sgx_default_qcnl.conf`

I set `use_secure_cerf = false` and this is what I got:
```
[QCNL] Data not found in local cache. Trying to fetch response from remote URL: 'https://api.trustedservices.intel.com/sgx/certification/v4/'. 

[QCNL] Request URL https://api.trustedservices.intel.com/sgx/certification/v4/pckcert?

[QCNL] HTTP status code: 404 

[QPL] No certificate data for this platform.

[get_platform_quote_cert_data ../td_ql_logic.cpp:302] Error returned from the p_sgx_get_quote_config API. 0xe011
tee_att_init_quote return 0x11001
tee_att_get_quote_size return 0x1100f
resp_size is 0
About to shutdown and close socket
erased a connection, now [0]
```

**Comment by mythi** [comment]
> I think I am missing something regarding the registration of the hw I am using.

Have you completed the registration?

**Comment by MatiasVara** [comment]
> > I think I am missing something regarding the registration of the hw I am using.
> 
> Have you completed the registration?

I just got my first quote. I had not completed the registration.

**Comment by mythi** [comment]
FYI: I've joined `#kubevirt-dev` Slack if you need to reach me quickly for any guidance.

**Comment by zhencliu** [comment]
/reopen

Matias and I are continuing the qgs support

**Comment by mythi** [comment]
@MatiasVara where can I help with the enhancement proposal?

**Comment by MatiasVara** [comment]
Hey @mythi, I am going to create a document and share it so we can collaborate there, WDYT?

**Comment by mythi** [comment]
Sounds good to me, thanks!

**Comment by MatiasVara** [comment]
> Sounds good to me, thanks!

This is the link to the document: https://docs.google.com/document/d/19pZ0J9yWZtKPYwyWduEErlYa_kL9tNDZyoXcJjz_vpk/edit?usp=sharing

**Comment by MatiasVara** [comment]
@mythi I am getting the following timeout recently, any idea? (sgx-common-2.25-4.el9.x86_64.rpm):
```
oc logs virt-handler-tf24r -n kubevirt -c qgs-launcher
+ set -e
+ export QCNL_CONF_PATH=/etc/qgs.conf
+ QCNL_CONF_PATH=/etc/qgs.conf
+ rm -rf /var/run/tdx-qgs/qgs.socket
+ exec /usr/sbin/qgs --debug --no-daemon --verbose
--no-daemon option found, will not run as a daemon
Use unix socket: /var/run/tdx-qgs/qgs.socket
Added signal handler
About to create QgsServer with num_thread = 0
About to start main loop
Added a new connection, now [1]
handle_read: status [system:0]
handle_read:[system:0Received 1024 bytes.]
process raw request [1024] bytes!.
unpack message successfully in thread [7f771b079740]
About to shutdown and close socket
timeout
erased a connection, now [0]
```

**Comment by mythi** [comment]
@MatiasVara it looks QGS tries to send something back to QEMU. Do you see any error logs on QEMU side for this?

**Comment by MatiasVara** [comment]
> [@MatiasVara](https://github.com/MatiasVara) it looks QGS tries to send something back to QEMU. Do you see any error logs on QEMU side for this?

Let me check.

**Comment by berrange** [comment]
> @MatiasVara it looks QGS tries to send something back to QEMU. Do you see any error logs on QEMU side for this?

I don't believe QGS has got as far as trying to send something back to QEMU - based on the logs it hasn't even processed the incoming request yet.

> handle_read:[system:0Received 1024 bytes.]
> process raw request [1024] bytes!.
> unpack message successfully in thread [7f771b079740]
> About to shutdown and close socket
> timeout
> erased a connection, now [0]

The ``unpack message successfully in thread`` line comes from the ``handle_raw_request`` method:

```
    void handle_raw_request() {
        std::ostringstream oss;
        oss << boost::this_thread::get_id();
        QGS_LOG_INFO("unpack message successfully in thread [%s]\n",
                     oss.str().c_str());
        asio::post(m_pool, [this, self = shared_from_this()] {
            boost::system::error_code ec;

            data_buffer resp = prepare_raw_response(const_cast<data_buffer &>(m_readbuf));
```

The `prepare_raw_response` method is trivial

```
    data_buffer prepare_raw_response(data_buffer const &req) {
        return get_raw_resp(req.data(), (uint32_t)req.size());
    }
```

and the `get_raw_resp` method should have emitted a log line very early:

```
    data_buffer get_raw_resp(const uint8_t *req, uint32_t req_size) {
        tee_att_error_t tee_att_ret = TEE_ATT_SUCCESS;
        data_buffer resp;

        if (ptr.get() == 0) {
            tee_att_error_t ret = TEE_ATT_SUCCESS;
            tee_att_config_t *p_ctx = NULL;
            QGS_LOG_INFO("call tee_att_create_context\n");
```

The log does NOT show `call tee_att_create_context` so it looks like `get_raw_resp` has not been invoked yet.

This suggests to me that ``asio:post`` has run, but the callback associated with it has not been triggered. 

A stack trace of QGS at the point where it has emitted ``unpack message successfully in thread`` but *before* it has triggered the timeout, might give an idea of what QGS is be doing when it is in this stuck state.

**Comment by mythi** [comment]
> I don't believe QGS has got as far as trying to send something back to QEMU - based on the logs it hasn't even processed the incoming request yet.

@berrange Right, apologies. I should have first checked this against a normal log which looks like:

```
...
unpack message successfully in thread [7a72c8233740]
call tee_att_create_context
create context in thread[7a72c8200640]
tee_att_init_quote return success
tee_att_get_quote_size return Success
tee_att_get_quote return Success
About to write response in thread [7a72c8200640]
About to shutdown and close socket
erased a connection, now [0]
```

> This suggests to me that `asio:post` has run, but the callback associated with it has not been triggered.

based on `About to create QgsServer with num_thread = 0` the thread pool is empty? Is there a value set for `number_threads = ` in `qgs.conf`?

**Comment by MatiasVara** [comment]
> > I don't believe QGS has got as far as trying to send something back to QEMU - based on the logs it hasn't even processed the incoming request yet.
> 
> [@berrange](https://github.com/berrange) Right, apologies. I should have first checked this against a normal log which looks like:
> 
> ```
> ...
> unpack message successfully in thread [7a72c8233740]
> call tee_att_create_context
> create context in thread[7a72c8200640]
> tee_att_init_quote return success
> tee_att_get_quote_size return Success
> tee_att_get_quote return Success
> About to write response in thread [7a72c8200640]
> About to shutdown and close socket
> erased a connection, now [0]
> ```
> 
> > This suggests to me that `asio:post` has run, but the callback associated with it has not been triggered.
> 
> based on `About to create QgsServer with num_thread = 0` the thread pool is empty? Is there a value set for `number_threads = ` in `qgs.conf`?

No, it is not.

**Comment by MatiasVara** [comment]
> > > I don't believe QGS has got as far as trying to send something back to QEMU - based on the logs it hasn't even processed the incoming request yet.
> > 
> > 
> > [@berrange](https://github.com/berrange) Right, apologies. I should have first checked this against a normal log which looks like:
> > ```
> > ...
> > unpack message successfully in thread [7a72c8233740]
> > call tee_att_create_context
> > create context in thread[7a72c8200640]
> > tee_att_init_quote return success
> > tee_att_get_quote_size return Success
> > tee_att_get_quote return Success
> > About to write response in thread [7a72c8200640]
> > About to shutdown and close socket
> > erased a connection, now [0]
> > ```
> > 
> > 
> >     
> >       
> >     
> > 
> >       
> >     
> > 
> >     
> >   
> > > This suggests to me that `asio:post` has run, but the callback associated with it has not been triggered.
> > 
> > 
> > based on `About to create QgsServer with num_thread = 0` the thread pool is empty? Is there a value set for `number_threads = ` in `qgs.conf`?
> 
> No, it is not.

@mythi Adding `-n=1` worked. Thanks.
